File: nomachinestatus.go
package main

import (
	"errors"
	"os"

	"github.com/mitchellh/go-ps"
)

const (
	ProcessNotFound   = -1
	serverProcessName = "nxserver.bin"
	clientProcessName = "nxexec"
)

// nomachineStatus represents the current state of NoMachine server and client
type nomachineStatus struct {
	HostName         string `json:"host_name,omitempty"`
	NoMachineRunning bool   `json:"no_machine_running,omitempty"`
	ClientAttached   bool   `json:"client_attached,omitempty"`
}

// IsActive returns true if NoMachine is running and has a client attached
func (s nomachineStatus) IsActive() bool {
	return s.NoMachineRunning && s.ClientAttached
}

func getFirstProcessByName(name string) (int, error) {
	if name == "" {
		return ProcessNotFound, errors.New("process name cannot be empty")
	}

	processes, err := ps.Processes()
	if err != nil {
		return ProcessNotFound, err
	}

	for _, process := range processes {
		if process.Executable() == name {
			return process.Pid(), nil
		}
	}

	return ProcessNotFound, errors.New("could not find process " + name)
}

// getStatus returns information about the NoMachine server and client status
func getStatus() (nomachineStatus, error) {
	hostName, err := os.Hostname()
	if err != nil {
		return nomachineStatus{}, err
	}

	status := nomachineStatus{HostName: hostName}

	noMachinePid, err := getFirstProcessByName(serverProcessName)
	if err != nil {
		// Consider logging the error
	}

	noMachineClientPid, err := getFirstProcessByName(clientProcessName)
	if err != nil {
		// Consider logging the error
	}

	if noMachinePid >= 0 {
		status.NoMachineRunning = true
	}
	if noMachineClientPid >= 0 {
		status.ClientAttached = true
	}

	return status, nil
}

File: Sources/App/nomachinestatus.swift
import Foundation
import Vapor

struct NoMachineStatus: Content {
    var hostName: String?
    var noMachineRunning: Bool = false
    var clientAttached: Bool = false
}

func getStatus() throws -> Result<NoMachineStatus, Error> {
    var status = NoMachineStatus()

    guard let hostName = Host.current().localizedName else {
        return .failure(
            NSError(
                domain: "", code: -1,
                userInfo: [NSLocalizedDescriptionKey: "Could not get hostname"]))
    }

    status.hostName = hostName

    let processes = try getRunningProcesses()
    if processes.contains(where: { $0.name == "nxserver.bin" }) {
        status.noMachineRunning = true
    }
    if processes.contains(where: { $0.name == "nxexec" }) {
        status.clientAttached = true
    }

    return .success(status)
}

struct ProcessResult: Content {
    var pid: Int
    var name: String
}

enum SysCtlError: Error {
    case FailedToGetProcessList1(message: String = "Failed to get process list step 1")
    case FailedToGetProcessList2(message: String = "Failed to get process list step 2")
}

/// Retrieves the list of currently running processes on the system.
/// - Returns: An array of `ProcessResult` containing the process ID and name.
/// - Throws: `SysCtlError` if the process list cannot be retrieved.
func getRunningProcesses() throws -> [ProcessResult] {
    var mib = [CTL_KERN, KERN_PROC, KERN_PROC_ALL]
    var size = 0

    guard sysctl(&mib, UInt32(mib.count), nil, &size, nil, 0) == 0 else {
        throw SysCtlError.FailedToGetProcessList1()
    }

    let entryCount = size / MemoryLayout<kinfo_proc>.stride
    let processList = UnsafeMutablePointer<kinfo_proc>.allocate(capacity: entryCount)
    defer { processList.deallocate() }

    guard sysctl(&mib, UInt32(mib.count), processList, &size, nil, 0) == 0 else {
        throw SysCtlError.FailedToGetProcessList2()
    }

    var processes = [ProcessResult]()

    for index in 0..<entryCount {
        var process = processList[index]
        let processId = process.kp_proc.p_pid
        if processId == 0 {
            continue
        }
        let processPcom = process.kp_proc.p_comm
        let name = withUnsafePointer(to: &process.kp_proc.p_comm) {
            $0.withMemoryRebound(
                to: CChar.self, capacity: MemoryLayout.size(ofValue: processPcom)
            ) {
                String(cString: $0)
            }
        }
        processes.append(ProcessResult(pid: Int(processId), name: name))
    }

    return processes
}
